pkg/cmd/pulumi/packagecmd/package_publish.go

```diff
@@ -236,3 +247,57 @@ func login(ctx context.Context, project *workspace.Project) (backend.Backend, er
 
 	return b, nil
 }
+
+// findReadme attempts to find a readme file in the given package source.
+// It tries to find a readme in the following order and returns the path to the first one it finds:
+// 1. The package source if it is a directory
+// 2. The installed plugin directory
+// If no readme is found, an empty string is returned.
+func (cmd *packagePublishCmd) findReadme(packageSrc string) (string, error) {
+	findReadmeInDir := func(dir string) string {
+		info, err := os.Stat(dir)
+		if err != nil && errors.Is(err, os.ErrNotExist) {
+			return ""
+		} else if err != nil {
+			return ""
+		}
+		if !info.IsDir() {
+			return ""
+		}
+		entries, err := os.ReadDir(dir)
+		if err == nil {
+			for _, entry := range entries {
+				if !entry.IsDir() {
+					name := strings.ToLower(entry.Name())
+					if name == "readme.md" {
+						return filepath.Join(dir, entry.Name())
+					}
+				}
+			}
+		}
+		return ""
+	}
+
+	// If the source is a directory, check if it contains a readme.
+	if readmeFromPackage := findReadmeInDir(packageSrc); readmeFromPackage != "" {
+		return readmeFromPackage, nil
+	}
+
+	// Otherwise, try to retrieve the readme from the installed plugin.
+	pluginSpec, err := workspace.NewPluginSpec(packageSrc, apitype.ResourcePlugin, nil, "", nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create plugin spec: %w", err)
+	}
+	pluginSpec.PluginDir = cmd.pluginDir
```

Comment:
Not sure if it's feasible or sensible, but part of me thinks we should try and create a plugin spec first thing when the command starts, because if we're trying to publish for a non-valid plugin then we've got issues? Then `findReadme` could take the `PluginSpec`, which'd simplify its logic and hone in the test cases to what we actually care about -- traversing the filesystem and finding the right thing, etc.

Reply:
We might not even have a plugin here (e.g. source just points to a json/yaml schema). But now that I'm thinking about that, I should make this case more explicit and not rely on `NewPluginSpec` to not error in that case (that's the current behavior)

Reply:
And regarding ensuring the plugin is valid first things first: we're doing that when extracting the schema. The schema is loaded from the plugin if it's not a file.